src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,538 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Security;
+using System.Threading;
+using System.Threading.Tasks;
+using CoreWCF.Channels;
+using CoreWCF.Configuration;
+using CoreWCF.Runtime;
+using CoreWCF.Security;
+
+namespace CoreWCF.Dispatcher
+{
+    /// <summary>
+    /// This is equivalent of SecurityChannelListener present in WCF codebase
+    /// </summary>
+    public class SecurityServiceDispatcher : IServiceDispatcher
+    {
+        private IRequestReplyCorrelator _requestReplyCorrelator;
+        private BindingContext bindingContext;
+        private ChannelBuilder channelBuilder;
+        private EndpointIdentity identity;
+        private SecurityBindingElement transportSecurityBinding;
+        private SecurityProtocolFactory securityProtocolFactory;
+        private SecuritySessionServerSettings sessionServerSettings;
+        private SecurityListenerSettingsLifetimeManager settingsLifetimeManager;
+        private bool sessionMode;
+        private bool sendUnsecuredFaults;
+        //ServiceChannelDispatcher to call SCT (just keep one instance)
+        private IServiceChannelDispatcher securityAuthServiceChannelDispatcher;
+        //ServiceChannelDispatcher to call real service (just keep one instance)
+        private IServiceChannelDispatcher innerServiceChanelDispatcher;
+        private IChannel outerChannel;
+        private Type acceptorChannelType;
+
+        public SecurityServiceDispatcher(SecurityBindingElement transportSecurityBindingElement, BindingContext context, IServiceDispatcher serviceDispatcher)
+        {
+            this.InnerServiceDispatcher = serviceDispatcher;
+            this.bindingContext = context;
+            this.transportSecurityBinding = transportSecurityBindingElement;
+            _requestReplyCorrelator = new RequestReplyCorrelator();
+            // this.securityProtocolFactory =  securityProtocolFactory; // we set it later from TransportSecurityBindingElement
+            //  this.settingsLifetimeManager = new SecurityListenerSettingsLifetimeManager(this.securityProtocolFactory, this.sessionServerSettings, this.sessionMode, this.InnerChannelListener);
```

Comment:
Can you open an issue to investigate whether SecurityListenerSettingsLifetimeManager needs to be reimplemented to handle service close

Reply:
I have already implemented that.. need clarifications